client/network/src/lib.rs

```diff
@@ -454,3 +454,106 @@ async fn wait_to_announce<Block: BlockT>(
 		);
 	}
 }
+
+/// Creates a new background task to wait for the relay chain to sync up and retrieve the parachain header
+pub async fn warp_sync_get<B, RCInterface>(
+	para_id: ParaId,
+	relay_chain_interface: RCInterface,
+	spawner: SpawnTaskHandle,
+) -> Result<oneshot::Receiver<<B as BlockT>::Header>, BoxedError>
+where
+	B: BlockT + 'static,
+	RCInterface: RelayChainInterface + 'static,
+{
+	let (sender, receiver) = oneshot::channel::<B::Header>();
+	spawner.spawn(
+		"cumulus-parachain-wait-for-target-block",
+		None,
+		async move {
+			tracing::debug!(
+				target: "cumulus-network",
+				"waiting for announce block in a background task...",
+			);
+
+			let _ = wait_for_target_block::<B, _>(sender, para_id, relay_chain_interface)
+				.await
+				.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						"Unable to determine parachain target block {:?}",
+						e
+					)
+				});
+		}
+		.boxed(),
+	);
+
+	Ok(receiver)
+}
+
+/// loops while waiting for the relay chain to sync, once complete return the latest parachain header to warp to
+async fn wait_for_target_block<B, RCInterface>(
+	sender: oneshot::Sender<<B as BlockT>::Header>,
+	para_id: ParaId,
+	relay_chain_interface: RCInterface,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+	B: BlockT + 'static,
+	RCInterface: RelayChainInterface + Send + 'static,
+{
+	let mut imported_blocks = relay_chain_interface.import_notification_stream().await?.fuse();
+	loop {
+		select! {
+			i = imported_blocks.next() => {
+					match i {
+					Some(_) => {
+						let is_syncing = relay_chain_interface
+						.is_major_syncing()
+						.await
+						.map_err(|e| {
+							tracing::error!(target: LOG_TARGET, "Unable to determine sync status. {e}")
+						})
+						.unwrap_or(false);
+
+						if !is_syncing {
+							let relay_chain_best_hash = relay_chain_interface
+								.last_finalised_block_hash()
+								.await
+								.map_err(|e| Box::new(e) as Box<_>)?;
+
+							let validation_data = relay_chain_interface
+								.persisted_validation_data(
+									relay_chain_best_hash,
+									para_id,
+									OccupiedCoreAssumption::TimedOut,
+								)
+								.await
+								.map_err(|e| Box::new(BlockAnnounceError(format!("{:?}", e))) as Box<_>)?
+								.ok_or_else(|| {
+									Box::new(BlockAnnounceError(
+										"Could not find parachain head in relay chain".into(),
+									)) as Box<_>
+								})?;
+
+							let target_block = B::Header::decode(&mut &validation_data.parent_head.0[..])
+								.map_err(|e| {
+									Box::new(BlockAnnounceError(format!(
+										"Failed to decode parachain head: {:?}",
+										e
+									))) as Box<_>
+								})?;
+
+							tracing::debug!(target: LOG_TARGET, "Target block reached {:?}", target_block);
+							let _ = sender.send(target_block);
+							return Ok(())
+						}
+						Delay::new(Duration::from_secs(30)).await;
```

Comment:
```suggestion
```

We actually don't need to sleep. When we have received a notification about an imported block, we should not doing major sync anymore and if not, the logic here isn't that problematic.